File: shared/schemas/audience.ts
/**
 * ReelLab shared contract — Audience.
 *
 * OWNER OF CHANGES: Developer 1 (AI / Simulation Lead), with review from
 * Developer 3 (frontend) and Developer 4 (backend).
 *
 * CONTRACT RULES (see docs/development-workflow.md):
 *  - Additive changes only. New fields must be optional.
 *  - Never rename or repurpose an existing field mid-hackathon.
 *  - Types only. This file must stay free of runtime code so every consumer
 *    can `import type` it with zero bundling or emit consequences.
 */

/** A probability or normalised score in the inclusive range 0..1. */
export type UnitScore = number;

/**
 * What the creator tells us before any AI runs.
 * This is the single input that seeds audience discovery.
 */
export interface AudienceRequest {
  /** Broad vertical, e.g. "fitness", "personal finance", "study with me". */
  niche: string;
  /** Primary audience in the creator's own words. */
  targetAudience: string;
  /** Optional secondary audience the creator also wants to reach. */
  secondaryAudience?: string;
  /** Geography, e.g. "India", "Tier-2 India", "US + Canada". */
  location: string;
  /** Content language, e.g. "English", "Hinglish". */
  language: string;
  /** What the creator is optimising for, e.g. "increase reach among beginners". */
  creatorGoal: string;
}

/**
 * One node of the discovered audience graph.
 *
 * Segments form a shallow tree: a segment with `parentSegment === null` is a
 * top-level niche; children are AI-discovered sub-niches.
 */
export interface AudienceSegment {
  id: string;
  name: string;
  description: string;
  /** `null` for a root niche, otherwise the parent segment's `id`. */
  parentSegment: string | null;
  /** Short descriptive traits, e.g. ["price-sensitive", "watches at night"]. */
  characteristics: string[];
  /** How relevant this segment is to the creator's goal, 0..1. */
  relevanceScore: UnitScore;
  /** Rough share of the reachable audience this segment represents, 0..1. */
  estimatedShare?: UnitScore;
  /** Free-form AI justification, surfaced in the UI for transparency. */
  rationale?: string;
}

/**
 * The full result of `POST /api/v1/audience/discover`.
 * `segments` is flat; the frontend rebuilds the tree from `parentSegment`.
 */
export interface AudienceGraph {
  /** Stable id so the frontend can cache and the backend can persist. */
  graphId: string;
  request: AudienceRequest;
  segments: AudienceSegment[];
  /** Edges between segments used by the propagation engine, if discovered. */
  adjacency?: SegmentAdjacency[];
}

/**
 * Directed likelihood that content spreads from one segment to another.
 * Consumed by the propagation engine; optional during early development.
 */
export interface SegmentAdjacency {
  fromSegmentId: string;
  toSegmentId: string;
  /** Probability that a share in `from` reaches `to`, 0..1. */
  spilloverProbability: UnitScore;
}

File: ai/schemas/audience.py
"""Pydantic mirror of `shared/schemas/audience.ts`.

Keep the two field-for-field compatible. If you change one, change the other in
the same pull request.
"""

from __future__ import annotations

from pydantic import Field

from .base import ReelLabModel


class AudienceRequest(ReelLabModel):
    niche: str
    target_audience: str
    secondary_audience: str | None = None
    location: str
    language: str
    creator_goal: str


class AudienceSegment(ReelLabModel):
    id: str
    name: str
    description: str
    parent_segment: str | None = None
    characteristics: list[str] = Field(default_factory=list)
    relevance_score: float = Field(ge=0.0, le=1.0)
    estimated_share: float | None = Field(default=None, ge=0.0, le=1.0)
    rationale: str | None = None


class SegmentAdjacency(ReelLabModel):
    from_segment_id: str
    to_segment_id: str
    spillover_probability: float = Field(ge=0.0, le=1.0)


class AudienceGraph(ReelLabModel):
    graph_id: str
    request: AudienceRequest
    segments: list[AudienceSegment]
    adjacency: list[SegmentAdjacency] = Field(default_factory=list)
